sklearn/utils/validation.py

```diff
@@ -155,6 +156,34 @@ def _shape_repr(shape):
     return "(%s)" % joined
 
 
+def check_memory(memory):
+    """Check that the memory is an instance of joblib.Memory.
```

Comment:
I think we need to come up with a better wording here and everywhere else in this PR. `joblib.Memory` is misleading since `joblib` may not be installed.

I would use "memory-like" (maybe Memory-like i.e.with a capital M), similarly to array-like in numpy.

The docstring would read something like

```py
"""Check that ``memory`` is memory-like.

memory-like means that ``memory`` can be converted into a
sklearn.externals.joblib.Memory instance (typically a str denoting the
``cachedir``) or has the same interface.

Parameters
----------
memory : memory-like or None

Returns
------
memory : object with the sklearn.externals.joblib.Memory interface

Raises
------
ValueError
    if ``memory`` is not memory-like
"""
```

And then use memory-like as the type info in all the docstrings, possibly with a rst link to check_memory for more details.